Introduce support for ARM's SCMI

LaurentiuM1234 opened the pull request:
**Design**
The SCMI support is comprised of three entities: the core, the transport and the protocols. Each of these is detailed below.

**Transport**
This layer is in charge of managing the channels and sending/receiving messages. A channel is a medium through which an SCMI protocol can send and receive messages.

Before being able to receive/send messages, a protocol needs to have a channel allocated/bound to it. There are two ways in which this can be done:

1. Dynamically - during the protocol setup phase, the SCMI core will attempt to request the transport layer driver a channel for each of the supported protocols. A protocol is considered "supported" if it has an associated "scmi_protocol" structure in the scmi_protocol linker section. The core will attempt to request a channel dynamically only if the transport layer doesn't support static channels.
2. Statically - if the transport layer supports it (signaled via the `ARM_SCMI_TRANSPORT_HAS_STATIC_CHANNELS` configuration), each protocol will have its channels assigned at compilation time. However, this means that the transport layer needs to use the following channel allocation scheme: if the transport-specific macro `DT_SCMI_TRANSPORT_PROTO_HAS_CHAN` evaluates to true on a protocol node that means the protocol will use a protocol-specific channel optionally described in the protocol devicetree node. If the macro evaluates to false then the protocol will use the base protocol channel which corresponds to the SCMI root node.

Let's assume we have the following devicetree:
```
firmware {
    scmi {
        compatible = "arm,scmi"; // select mailbox-based transport
        shmem = <&myshmem1>;

        protocol@14 {
            compatible = "arm,scmi-clock";
            shmem = <&myshmem2>;
            #clock-cells = <1>;
        };
    };
};
```

and the following definition for `DT_SCMI_TRANSPORT_PROTO_HAS_CHAN`:

```
#define DT_SCMI_TRANSPORT_PROTO_HAS_CHAN(node_id, idx)\
             DT_PROP_HAS_IDX(node_id, shmem, idx)
```

First of all, the devicetree snippet indicates that we're dealing with two protocols: base (represented by the `scmi` node) and clock (represented by the `protocol@14` node). Applying `DT_SCMI_TRANSPORT_PROTO_HAS_CHAN` with idx 0 (i.e: the transport channel) to the `protocol@14` node will result in the macro evaluating to true. This means that in the static allocation scheme, the clock protocol will use its own channel separate from the one used by the base protocol.

On the contrary, in the following devicetree snippet:
```
firmware {
    scmi {
        compatible = "arm,scmi"; // select mailbox-based transport
        shmem = <&myshmem1>;

        protocol@14 {
            compatible = "arm,scmi-clock";
            #clock-cells = <1>;
        };
    };
};
```

the clock protocol would use base's TX channel as the `DT_SCMI_TRANSPORT_PROTO_HAS_CHAN` would evaluate to false when applied to `protocol@14`.

**For static channels, the transport layer is responsible for validating the information used to build the channels passed through the devicetree at compilation time**.


**Protocols**
The protocol layer is in charge of preparing the data and calling into the transport layer to send/receive messages. The data to be sent/received is encapsulated in a `struct scmi_message`. To send a message, all that a protocol has to do is call `scmi_send_message()`. This function will also wait for the reply so it's a blocking function.

**Core**
This is a glue layer between the transport and the protocols. What this layer does is it:

1. Makes sure a channel can only be accessed by a single protocol at a time.
2. Binds channels to protocols and initializes the channels.
3. Changes the send method based on the initialization phase of the kernel. Since in `PRE_KERNEL_1` or `PRE_KERNEL_2` the kernel primitives are unusable, the core will have to use a different method to wait for a reply to a message (i.e: polling). In `POST_KERNEL`, the `scmi_send_message()` function is notified that the reply has arrived via a binary semaphore. Also, since there's no concurrency in `PRE_KERNEL_1` and `PRE_KERNEL_2` phases there's really no need to use a mechanism to make sure only one protocol accesses a channel at a certain time.

**Porting guide**
**Transport layer driver**
To write a transport layer driver, you'll have to introduce a new compatible for the `scmi` node. After doing so, your driver will have to implement the operations from `struct scmi_transport_api`. All operations are mandatory except for `init` and `request_channel` (which will become mandatory if your transport layer driver doesn't support static channel allocation). The `init` function is basically the same as the init function you'd pass to macros such as `DEVICE_DT_DEFINE` and `DEVICE_DT_INST_DEFINE`. This, in turn, will be the first function called by `scmi_core_transport_init()`.

After implementing the required transport operations, you'll have to call `DT_INST_SCMI_TRANSPORT_DEFINE()`, which is nothing more than a wrapper over `DEVICE_DT_INST_DEFINE()`. The reason for this is because the SCMI core needs to have its `scmi_core_transport_init()` function called when the transport layer driver is initialized to perform channel binding and setup for the protocols.

**Protocol driver**
This is relatively straightforward. All you have to do is call `DT_INST_SCMI_PROTOCOL_DEFINE()` in your driver as you'd normally with `DEVICE_DT_INST_DEFINE` and `DEVICE_DT_DEFINE`. For protocol drivers which don't really need to have a `struct device` created for it (like it's the case for pinctrl) there's also the `DT_SCMI_PROTOCOL_DEFINE_NODEV` variant. Note that using this variant also means you won't have an init function.

**What the initial support brings to the table**
1. Clock control operations: gating, ungating, and rate query
2. Pinctrl operations: pin configuration
3. Mailbox-based transport

**What the initial support lacks**

1. Support for notifications
2. Support for base protocol (and the other protocols not mentioned)
4. Support for delayed replies

**TODOs**

1. ~This draft has no documentation. This will be addressed later on depending on the feedback of the community. The main point of this draft is to get some feedback on the design.~

dbaluta commented on drivers/clock_control/CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_RENESAS_RA          clock_cont
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_PWM                 clock_control_pwm.c)
 zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_RPI_PICO            clock_control_rpi_pico.c)
+zephyr_library_sources_ifdef(CONFIG_CLOCK_CONTROL_ARM_SCMI            clock_control_arm_scmi.c)
 
```
Looks like at some point people tried to kept this sorted but there were some slip-ups. Can you somehow add this just before CONFIG_CLOCK_CONTROL_ADSP

dbaluta commented on drivers/firmware/scmi/core.c:
```diff
@@ -0,0 +1,223 @@
+#define SCMI_CHAN_SEM_TIMEOUT_USEC 500
+
+int scmi_status_to_linux(int scmi_status)
+{
```
These are not Linux specific errors. Lets go with `scmi_status_to_errno`.

dbaluta commented on drivers/firmware/scmi/shmem.c:
```diff
@@ -0,0 +1,201 @@
+}
+
+
+int scmi_shmem_write_message(const struct device *shmem, struct scmi_message *msg)
```
no double empty lines please.

LaurentiuM1234 commented:
**V2 changes**

- fixed compliance failure
- fixed some function prototype alignment issues signaled by checkpatch
- addressed Daniel's initial comments. Mention: `CONFIG_CLOCK_CONTROL_ARM_SCMI` was added **after** `CONFIG_CLOCK_CONTROL_ADSP` in `drivers/clock_control/CMakeLists.txt`

dleach02 commented:
@ithinuel can you provide some review support to this RFC?

yangbolu1991 commented on drivers/firmware/scmi/shmem.c:
```diff
@@ -0,0 +1,200 @@
+	/* mismatch between expected reply size and actual size? */
+	if (msg->len != (layout->len - sizeof(layout->msg_hdr))) {
+		LOG_ERR("bad message len. Expected 0x%x, got 0x%x",
+			msg->len, (layout->len - sizeof(layout->msg_hdr)));
```
A waring here when build for A55 .

```
zephyr/drivers/firmware/scmi/shmem.c:86:25: warning: format '%x' expects argument of type 'unsigned int', but argument 3 has type 'long unsigned int' [-Wformat=]
   86 |                 LOG_ERR("bad message len. Expected 0x%x, got 0x%x",
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   87 |                         msg->len, (layout->len - sizeof(layout->msg_hdr)));
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                |
      |                                                long unsigned int
```

yangbolu1991 commented on drivers/firmware/scmi/Kconfig:
```diff
@@ -0,0 +1,52 @@
+
+config ARM_SCMI_SHMEM_INIT_PRIORITY
+	int "SCMI shared memory (SHMEM) initialization priority"
+	default 1
```
Should we increase the priority value, for example, 4?
Before the SHMEM, there may be other dependencies. For example,
- MU (hardware message unit)
- GIC controller which MU may use.
Thanks.

LaurentiuM1234 replied:
Good question. This is indeed an issue so it might be worth opening a discussion. Although, the SHMEM component is not the problem here as it doesn't have a dependency. The problematic components are the transport layer driver and the core (since this is initialized together with the transport layer).

For this particular case (i.e: 95) we're going to end up having to modify the initialization priorities of **some** components (SCMI or non-SCMI components) no matter what priorities we choose for the SCMI components (i.e: SHMEM and TRANSPORT).

As such, I agree with your proposal to increase the priority number (i.e: decrease the init priority) to allow for a bit more leeway when configuring the priorities of other components. This should be done for both the `SHMEM` and `TRANSPORT` components even though `SHMEM` has no dependency but should ideally have the same initialization priority as `MBOX` as they both make up the SCMI channels.

yangbolu1991 commented on drivers/clock_control/clock_control_arm_scmi.c:
```diff
@@ -0,0 +1,102 @@
+	.on = scmi_clock_on,
+	.off = scmi_clock_off,
+	.get_rate = scmi_clock_get_rate,
+};
```
Is there any plan to implement set_rate? I think that may be much complex to consider parent clock.
But that should be requirement.
Thanks.

LaurentiuM1234 replied:
There was no plan to add this in the initial support. I tested this on 95 with some basic samples (i.e: `hello_world` and `synchronization`) in which `set_rate()` was not needed so this is why this wasn't included here. Of course, this feature is important so it should be added at some point. Do we need to have this feature in the initial support or is it fine if we add it later on?

yangbolu1991 replied:
Sure, it's ok for me to have it added in the future :)

yangbolu1991 commented:
BTW, I verified the pull request on i.MX95 both A core and M core. All worked fine.
Thanks.

LaurentiuM1234 commented:
**V3 changes**

- Fixed compilation warning pointed out by @yangbolu1991 (thanks!)
- Changed `ARM_SCMI_TRANSPORT_INIT_PRIORITY` to 20
- Changed `ARM_SCMI_SHMEM_INIT_PRIORITY` to 15

The priority number changes should allow for a bit more leeway in cases where we need to update the init priorities of the dependencies.

LaurentiuM1234 commented:
PR has been out for a while now, dropping `RFC` tag.

JiafeiPan commented on include/zephyr/drivers/firmware/scmi/clk.h:
```diff
@@ -0,0 +1,96 @@
+ */
+enum scmi_clock_message {
+	SCMI_CLK_MSG_PROTOCOL_VERSION = 0x0,
+	SCMI_CLK_MSG_NEGOTIATE_PROTOCOL_VERSION = 0x10,
```
Put protocol 0x10 to be after 0xf?

LaurentiuM1234 replied:
thanks, will have these ordered by ID value

JiafeiPan commented on include/zephyr/drivers/firmware/scmi/util.h:
```diff
@@ -0,0 +1,281 @@
+#define SCMI_PROTOCOL_RESET_DOMAIN 22
+#define SCMI_PROTOCOL_VOLTAGE_DOMAIN 23
+#define SCMI_PROTOCOL_PCAP_MONITOR 24
+#define SCMI_PROTOCOL_PINCTRL 25
```
May I know whether SCMI protocol IDs are unified between different vendors or platforms?

LaurentiuM1234 replied:
Depends on the range. If we're talking about the 0x10 - 0x19 range then yes. For instance, 0x10 will always be base protocol and so on.

JiafeiPan reviewed (commented):
looks fine for me with some minimal comments/questions.

LaurentiuM1234 commented:
**V4 changes**

- addressed @JiafeiPan's comments

decsny commented:
I think this needs to be presented to the architecture working group especially considering the impending clocks rework that has already been discussed multiple times in that forum

@carlescufi should the label be added?

danieldegrasse commented:
Won't be able to join the Arch WG,  but I think the core question from me would be if SCMI will map well into the proposed clock management framework- from what I can tell setting/getting rates would map well, but I am not sure how well gating/ungating clocks would. Perhaps a clock is automatically gated if the rate is set to 0?

LaurentiuM1234 commented:
**V5 changes**

1. Drop usage of `scmi_channel` linker section. It's whole point was to allow for iterating through the SCMI channels but since we don't need this we can drop it. Channels are still placed in RAM.
2. Simplify channel binding code in `scmi_core_protocol_setup`.
3. Adjust `help` message of `ARM_SCMI_TRANSPORT_HAS_STATIC_CHANNELS`. Clearly state what the default static channel allocation scheme is.

dbaluta commented:
For anyone interested we will be joining today's Architecture Group Meeting.

https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group

dbaluta commented:
@decsny @dleach02  looks like we need to update the docs because it is not enough just to add Architecture Review label you need to announce yourself on discord.

Will try to schedule the review for this for next Tuesday. Sorry guys :(

![image](https://github.com/user-attachments/assets/8f0b5d82-2fc1-49ec-8e57-e1f473d68a20)

fabiobaltieri commented:
Arch wg notes:
- Laurentiu presented the topic
- @bjarki-andreasen and @fabiobaltieri asked about the location of this code `drivers/firmware`, this matches the Linux path `drivers/firmware/arm_scmi`
- @fabiobaltieri asked about using devicetree entries for protocol definition, that is also matching how it's done in Linux

iuliana-prodan reviewed (dismissed):
LGTM

bjarki-andreasen reviewed (changes requested):
I would like to see this PR split into multiple commits, one for adding the scmi, one for adding the clock_control device driver. And finally, a sample or other implementation showing in-tree usage of SCMI, specifically getting a struct device handle for the clock controller instance for example :) I heard there was a separate PR which did this, please add that as a commit to this PR so we get the whole picture :)

dbaluta commented:
Here is the PR that adds Zephyr support for imx95 soc with M7 and a55 cores. 

https://github.com/zephyrproject-rtos/zephyr/pull/74920

Example of the pinctrl scmi driver: 
- https://github.com/zephyrproject-rtos/zephyr/pull/74920/commits/320da60c6226334299e987cc50af84b2b24cfc7b

Example of how to describe the scmi node in dts:

https://github.com/zephyrproject-rtos/zephyr/pull/74920/commits/e6b8fbf7d1cb125c518b2de7d1be1de129a118de#diff-77b68905139384f01324b12bca383dec7671af1250c5d2f1ab1e0056412ef57bR41

LaurentiuM1234 commented:
**V6 changes**

`dts/bindings/firmware/arm,scmi.yaml`:

1. Clearly state that this binding is for shared memory and doorbell-based transport.
2. Add devicetree example
3. `shmem`, `mboxes`, and `mbox-names` are now all marked as required.
4. Add description for all properties

`dts/bindings/firmware/arm,scmi-clock.yaml`:

1. Add description for `clock-cells` property.

`drivers/firmware/scmi/mailbox.h`:

1. Add more `BUILD_ASSERT`-based statements for checking if channel properties are encoded properly in DTS.
2. TX channel mailbox is now fetched using `_SCMI_MBOX_CHAN_DBELL` (old way: using `MBOX_DT_SPEC_GET` directly).  This makes the error messages easier to read whenever users forget to add "tx" mbox (now it's just the failed build assert).


Misc:

1. PR has been split into 3 commits: core support, pinctrl, and clock management.
2. Rebased


> And finally, a sample or other implementation showing in-tree usage of SCMI, specifically getting a struct device handle for the clock controller instance for example :) I heard there was a separate PR which did this, please add that as a commit to this PR so we get the whole picture :)

Very reasonable. The main concern I have with including https://github.com/zephyrproject-rtos/zephyr/pull/74920 in this PR is that it will make reviewing this harder as said PR doesn't contain just SCMI-related additions, but the whole platform support. Also, there's no changes to the way "struct device" for the clock controller is obtained for SCMI.

As for the sample: that would be good to have actually. My plan was to add a "hello world" sample using base protocol but this would require having a platform that uses SCMI. We do have https://github.com/zephyrproject-rtos/zephyr/pull/74920 but that depends on this PR.

bjarki-andreasen reviewed (commented):
There are some files which belong to incorrect commits, like `drivers/clock_control/CMakeLists.txt` being changed in the first commit :)

Regarding the sample/implementation: Given that #74920 does implement SCMI, I think it is okay to merge this PR even though there technically isn't an upstream implementation yet, given the size of #74920, which is the upstream implementation. #74920 itself is "sample enough" I think. By sample I just mean some upstream implementation, not necessarily a sample like hello_world :)

LaurentiuM1234 commented:
**V7 changes**

1. Moved `drivers/clock_control/CMakeLists.txt` to commit introducing clock protocol support.
2. Changed commit message for commit introducing clock protocol support. Previously, it didn't mention anything about introducing clock controller driver and DT binding.


> There are some files which belong to incorrect commits, like `drivers/clock_control/CMakeLists.txt` being changed in the first commit :)

thanks a lot! should be fixed now.

bjarki-andreasen commented on drivers/firmware/Kconfig:
```diff
@@ -0,0 +1,16 @@
+
+config ARM_SCMI
+	bool "Support for ARM's SCMI"
+	default n
```
`default n` is already the default, please remove it :)

bjarki-andreasen commented on drivers/firmware/scmi/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+# SCMI core files
+zephyr_library_sources_ifdef(CONFIG_ARM_SCMI core.c)
+zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_MAILBOX_TRANSPORT mailbox.c)
+zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_MAILBOX_TRANSPORT shmem.c)
```
This should either be split into two configs; one for mailbox, one for shmem, or, combine them like
```
zephyr_library_sources_ifdef(CONFIG_ARM_SCMI_MAILBOX_TRANSPORT mailbox.c shmem.c)
```

bjarki-andreasen replied:
edit, please do split them, they both have separate init priorities, so they are seemingly separate features :) if they depend on each other, use Kconfig to manage this

bjarki-andreasen commented on drivers/firmware/scmi/Kconfig:
```diff
@@ -0,0 +1,38 @@
+
+config ARM_SCMI_TRANSPORT_HAS_STATIC_CHANNELS
+	bool "Transport layer has static channels"
+	default n
```
please remove redunant `default n`

bjarki-andreasen commented on include/zephyr/linker/common-ram.ld:
```diff
@@ -12,6 +12,10 @@
 #endif
 #endif /* NETWORKING */
 
+#ifdef CONFIG_ARM_SCMI
+ITERABLE_SECTION_RAM(scmi_protocol, Z_LINK_ITERABLE_SUBALIGN)
+#endif /* CONFIG_ARM_SCMI */
+
```
An entry in `zephyr/cmake/linker_script/common/common-rom.cmake` is required as well

LaurentiuM1234 replied:
typo? did you mean `common-ram.cmake` by any chance?

bjarki-andreasen replied:
yes, RAM, sorry :)

LaurentiuM1234 committed:
firmware: introduce SCMI core support

Introduce core support for ARM's SCMI (System Control and
Management Interface). This includes:

* shared memory (SHMEM) driver. This consists of a suite
of functions used to interact with the shared memory area.

* shared memory and doorbell-based transport layer driver.
Data is passed between platform and agent via shared
memory. Signaling is done using polling (PRE_KERNEL) and
doorbells (POST_KERNEL). This makes use of Zephyr MBOX API
(for signaling purposes) and the SHMEM driver (for polling
and data transfer).

* core driver - acts as glue between transport and protocol
layers. Provides synchronized access to transport layer
channels and channel assignment/initialization.

* infrastructure for creating SCMI protocols

This is based on ARM's SCMI Platform Design Document: DEN0056E.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
firmware: scmi: add support for clock management protocol

This includes:
	1) Source containing helper functions, each
	implementing a command from the clock management
	protocol.

	2) A clock controller driver making use of said
	helper functions and implementing the clock
	subsystem API.

	3) A DT binding for clock protocol node.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
firmware: scmi: add support for pinctrl protocol

This includes helper function for pin configuration
and a DT binding for the pinctrl DT node.

There's two important notes to be made regarding this
protocol:

* pinctrl drivers have no subsytem API to implement as opposed
to clock control drivers. Because of this (and the fact that
`pinctrl_configure_pins()` doesn't require a `struct device`
handle) the pinctrl driver consists only of a helper function,
which implements the `PINCTRL_CONFIGURE_PINS` command.
Additionally, the `scmi_protocol` structure is defined inside
the pinctrl helpers source file to avoid redundant code
(otherwise, each SCMI-based pinctrl driver would have to define
it its source file).

* each vendor may have their own set of pin propeties and DT
representations for them. Because of this, there can't be a
generic, SCMI-based pinctrl driver. As such, each vendor who
wants to use the SCMI support for pinctrl operations will have
to implement their pinctrl driver (which, to put it simply,
revolves around implemeting `pinctrl_configure_pins()`) and
make use of the pin configuration function introduced in this
commit. Moreover, this means that each vendor will have control
over the way their pin properties are encoded in the
`scmi_pinctrl_settings` structure.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented:
**V8 changes**

`drivers/clock_control/Kconfig.arm_scmi`:

1. Remove all `select` and `depends on` statements from `CLOCK_CONTROL_ARM_SCMI` except for the `depends on ARM_SCMI_CLK_HELPERS` statement. This is enough to assure that `ARM_SCMI` is selected and the clock protocol DT node is enabled.

Misc:

1. Add entry for `scmi_protocol` iterable section in `common-ram.cmake`.
2. Drop `default n` statements indicated by @bjarki-andreasen 
3. Add configuration for shmem driver: `CONFIG_ARM_SCMI_SHMEM`. Additionally, handle dependency between the transport driver and the shmem driver via Kconfig `depends on`.

@bjarki-andreasen could you please take another look? Comments should be addressed now. Thanks!

bjarki-andreasen reviewed (dismissed):
Nice work!

nashif commented:
This is missing very basic documentation

dbaluta commented:
> This is missing very basic documentation

@nashif do you have any suggestion where should we place the doc file? There is no clear file path that comes to our mind right now.

bjarki-andreasen commented:
@dbaluta I will propose `doc/hardware/arch/arm-scmi.rst`

LaurentiuM1234 committed:
doc: add SCMI documentation

Add documentation page for SCMI.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented:
**V9 changes**

1. Add documentation page under `doc/hardware/arch/arm-scmi.rst` as suggested by @bjarki-andreasen.

LaurentiuM1234 commented:
@nashif @henrikbrixandersen @fabiobaltieri anything left to address? if not, can we proceed with merge?

bjarki-andreasen commented:
I removed the arch meeting label, I don't think there is anything to discuss, should be ready merge now :)
